Initial skeleton for profile loading support

leon-gh opened the pull request:
Currently just a placeholder for selecting a filename. Debug log is displayed when action is invoked. Next step is to integrate it with ui.backend to perform the actual profile loading.

leon-gh committed:
Load profile support

* Load another profile as an action.

WhiteMagic commented on action_plugins/load_profile/__init__.py:
```diff
@@ -0,0 +1,165 @@
+        super().__init__(behavior_type)
+
+        # Model variables
+        self.profile_filename = ""
```
Might be worth to use a `pathlib.Path` object here so you can easily verify the provided string is valid and possibly also use that for the valid aspect. The `is_valid()` bit is mainly there to avoid saving "bad" configuration bits.

leon-gh replied:
I implemented is_valid() using os.isfile and os.access. If you prefer pathlib.Path() then the profile_filename will become None at init and object when assigned. Let me know.

WhiteMagic replied:
I think both are fine, in most of the new parts of Gremlin I've starting using `Path` objects instead of strings and `os.` calls.

leon-gh replied:
OK, then I'll make it consistent, no issues.

leon-gh replied:
Turns out pathlib.Path code lacks .access() to validate whether file is accessible. Left os methods since Path code turned out more lines and more complex.

WhiteMagic commented on action_plugins/load_profile/__init__.py:
```diff
@@ -0,0 +1,165 @@
+    properties = [
+        ActionProperty.ActivateOnBoth
+    ]
+    input_types = [
```
This should only be for `JoystickButton` and `Keyboard` as they are the only two "push" style inputs. The other two inputs will show this action dynamically when they are switched to behave like that via the UI.

leon-gh replied:
OK. What about JoystickHat? Is it not a push style input?

WhiteMagic replied:
A hat reports 9 values, directions & center, but only gives you a "new value" not a "let go of this direction". One can map that onto button-behavior which Gremlin does when the hat is treated as a button in Gremlin. So actions that require a "press" action only make sense for the two button-like inputs though others can be made to behave as such.

leon-gh replied:
OK, understood.

WhiteMagic commented on action_plugins/load_profile/__init__.py:
```diff
@@ -0,0 +1,165 @@
+    functor = LoadProfileFunctor
+    model = LoadProfileModel
+
+    properties = [
```
You probably don't want this, as that would trigger a profile load when the input is pressed and again when released. The better one is `ActivateOnPress` as that only is the default behavior not the one that it is restricted to. You also may want to add the `AlwaysExecute` property as otherwise you won't be able to load a profile when the current profile is paused. In R14 that "run when paused" is a per-action-class setting rather than per-instance as was the case in R13.

leon-gh replied:
Noted.

WhiteMagic commented on action_plugins/load_profile/__init__.py:
```diff
@@ -0,0 +1,165 @@
+    ) -> None:
+
+        if not self._should_execute(value):
+                return
```
That indentation seems too deep

leon-gh replied:
Will fix the indent and other unused imports as a part of cleanup.

WhiteMagic commented on action_plugins/load_profile/__init__.py:
```diff
@@ -0,0 +1,165 @@
+    def __init__(self, action: LoadProfileData):
+        super().__init__(action)
+
+        self._profile_filename = action.profile_filename
```
No real need to store this, as everything is stored in the data object which is stored as `self.data`. This has the nice side-effect that changing of values while the profile is active are automatically reflected in the running profile.

leon-gh replied:
OK.

WhiteMagic commented on action_plugins/load_profile/LoadProfileAction.qml:
```diff
@@ -0,0 +1,77 @@
+
+   }
+
+   FileDialog {
```
Technically this should need `import Qt.labs.platform` but maybe through the way that the UI is constructed this is implicitly available.

WhiteMagic replied:
Would be a good idea to make use of the filter function of this to only show files that should be loadable, i.e. XML files.

https://doc.qt.io/qt-6/qml-qt-labs-platform-filedialog.html#nameFilters-prop

leon-gh replied:
FileDialog needs import QtQuick.Dialogs. Added filters for *.xml (had plans, but just didn't make it in yet :-))

WhiteMagic commented on action_plugins/load_profile/LoadProfileAction.qml:
```diff
@@ -0,0 +1,77 @@
+        id: fileDialog
+        title: "Select a File"
+        onAccepted: {
+            _profile_filename.text = new URL(selectedFile).pathname.substring(1)
```
- Might also just use `file` here as that's the value of the path selected upon acceptance while `selectedFile` is constantly changing.
- Does this return the complete absolute path to the selected string?

leon-gh replied:
FileDialog returns selectedFile as url (file:///C:/path/file). 
I converted it to string, then used .substring("file:///".length)) on it. If not OK, please advise how do you want me to change it.

WhiteMagic replied:
As long as you get the valid absolute path out of it I would say that's what you want. Was mostly wondering, as I don't remember anymore what those calls spat out.

leon-gh replied:
Got it, n/p.

WhiteMagic commented:
Left you a few comments but otherwise looks like there is nothing that seriously is broken with it. There is also a `qml_notes.md` document in the root directory which contains a bunch of things I've learned about QML and its quirks. This book is also somewhat useful to get an idea of some of QML's aspects https://www.qt.io/product/qt6/qml-book

leon-gh commented:
> Left you a few comments but otherwise looks like there is nothing that seriously is broken with it. There is also a `qml_notes.md` document in the root directory which contains a bunch of things I've learned about QML and its quirks. This book is also somewhat useful to get an idea of some of QML's aspects https://www.qt.io/product/qt6/qml-book

Yeah, went through this book by looking at what I needed :-). What I miss is ability to debug qml somehow. If you can debug QML it would be nice to have the same IDE configurations.

leon-gh committed:
Code review changes and cleanup

* qml cleanup:
   - using toString instead of URL object in FileDialog
   - removed _ from TextField id
* addressed code review changes
* flake8 fixes

WhiteMagic commented:
Yeah ... debugging QML is not something I ever figured out if you can really do. It's really all kinds of trying things out and making stuff visual, so the worst kind of printf debugging.

WhiteMagic commented on action_plugins/load_profile/LoadProfileAction.qml:
```diff
@@ -0,0 +1,77 @@
+        }
+
+        TextField {
+            id: profile_filename
```
Just as an FYI I tend to do `_camelCase` for IDs (semi-successful at remembering this myself) as Qt and QML use that which is also why you'll see slots use camel case for method names. And I add the underscore so I can easily recognize an ID reference in some of the bigger QML files.

leon-gh replied:
OK, will update.

WhiteMagic commented on action_plugins/load_profile/__init__.py:
```diff
@@ -0,0 +1,171 @@
+            f"Loading profile ...{self.data.profile_filename}"
+        )
+
+        be = backend.Backend()
```
Maybe worth checking the file is still valid here or in the `_from_xml()` (likely better since I hope people don't move a profile when they actively use Gremlin) call to prevent attempting to load an inaccessible one. I think you should be fine with regards to profile errors as that error chain leads to a UI error dialogue. Starting can throw exceptions and they might be uncaught until termination possibly.

leon-gh replied:
Added validator to _from_xml()

leon-gh committed:
Addressed review comments

* QML identifiers use _camelCase
* LoadProfileData no longer accepts input from JoystickHat
* Raise exception if load-profile action is associated with non-existing
  or not accessible profile when the profile is loaded.

WhiteMagic commented on action_plugins/load_profile/__init__.py:
```diff
@@ -133,24 +132,27 @@ def __init__(
 
         # Model variables
-        self.profile_filename = ""
+        self._profileFilename = ""
```
Ah sorry I think I confused you, the only place I use non-pythonic camelCase variables is QML exposed things, in the data class and really everywhere else it's normal Python convention, i.e. `self.profile_name` for this one here unless you intend to expose it via a property.

leon-gh replied:
n/p, I thought *Data class was directly linked to QML, but it's via *Model getter/setters - new to Qt/QML.

leon-gh committed:
Review comments

* renamed _profileFilename to profile_name in python files.

WhiteMagic commented:
Awesome work, thanks.
